feat(pom.xml): add `dependsOn` and `Indirect` support

DmitriyLewen opened the pull request:
## Description
Add `ID` and `dependsOn` support.
Fill `Indirect` field.

Some nuances:
- root/module artifact is marked as `Direct`.
- dependencies of root/module artifact are marked as `Direct` (this is necessary to be able to exclude indirect dependencies from Trivy report - https://github.com/aquasecurity/trivy/discussions/4870).
- root artifact doesn't contains modules in `dependsOn` (see more [here](https://github.com/DmitriyLewen/go-dep-parser/blob/bb82a5149a74ed72accd9a8a171cb08e4378f15e/pkg/java/pom/parse_test.go#L661-L691))

## Related Issue
- Close #126
- aquasecurity/trivy/issues/3272

roi-orca commented:
@DmitriyLewen can we add the indication whether the dependency is direct or not using the `Indirect` attribute or it is an issue by itself?

DmitriyLewen commented:
Hello @roi-orca 
Good idea!
Let me check this.

DmitriyLewen commented:
@roi-orca i added Indirect field.

roi-orca commented:
@DmitriyLewen when a package will be treated as indirect in a pom.xml? for the dependencies of the dependencies?

DmitriyLewen commented:
> for the dependencies of the dependencies?

Right.

some example:
- rootPkg (GAV of scanned pom.xml file) - **Direct**
  - pkgA (pkg from `dependencies` tag) - **Direct**
    - pkgC - **Indirect**
      - pkgD - **Indirect**
  - pkgB - **Direct**

take a look this test - https://github.com/aquasecurity/go-dep-parser/blob/a0eb566217664c77625bf3de4201fe9da7c45099/pkg/java/pom/parse_test.go#L740

knqyf263 commented on pkg/java/pom/utils.go:
```diff
@@ -19,3 +20,7 @@ func isProperty(version string) bool {
 	return false
 }
+
+func PackageID(name, version string) string {
```
Do we need to export the function?

DmitriyLewen replied:
Fixed

knqyf263 reviewed (commented):
I pushed refactoring, but please correct me if I'm wrong.

DmitriyLewen commented on pkg/java/pom/parse.go:
```diff
@@ -172,22 +189,69 @@ func (p *parser) parseRoot(root artifact) ([]types.Library, []types.Dependency,
+			// mvn only takes top dependencies
+			// this is needed to reproduce mvn logic
+			// take a look at `soft requirement`, `soft requirement with transitive dependencies`
+			// and `hard requirement for the specified version` tests
```
It makes sense for SBOM.
But i thought about versions for these dependencies ( i mean cases when version was overwritten).

I checked [maven Cyclonedx plugin](https://github.com/CycloneDX/cyclonedx-maven-plugin):
pom.xml dependencies:
```
    <dependencies>
       <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-databind</artifactId>
            <version>2.15.0</version>
        </dependency>
        <dependency>
            <groupId>com.fasterxml.jackson.core</groupId>
            <artifactId>jackson-core</artifactId>
            <version>2.15.2</version>
        </dependency>
    </dependencies>
```
maven dep tree:
```
[INFO] com.example:example:jar:1.0.0
[INFO] +- com.fasterxml.jackson.core:jackson-databind:jar:2.15.0:compile
[INFO] |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.15.0:compile
[INFO] \- com.fasterxml.jackson.core:jackson-core:jar:2.15.2:compile
```
dependencies from built (https://github.com/CycloneDX/cyclonedx-maven-plugin) SBOM file:
```
  "dependencies" : [
    {
      "ref" : "pkg:maven/com.example/example@1.0.0?type=jar",
      "dependsOn" : [
        "pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.15.0?type=jar",
        "pkg:maven/com.fasterxml.jackson.core/jackson-core@2.15.2?type=jar"
      ]
    },
    {
      "ref" : "pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.15.0?type=jar",
      "dependsOn" : [
        "pkg:maven/com.fasterxml.jackson.core/jackson-annotations@2.15.0?type=jar",
        "pkg:maven/com.fasterxml.jackson.core/jackson-core@2.15.2?type=jar"
      ]
    },
```

So we can use similar logic. 
I will update PR.

knqyf263 replied:
Yeah, the cyclonedx plugin behaves as I suggested. `com.fasterxml.jackson.core:jackson-annotations:2.15.0` => `com.fasterxml.jackson.core:jackson-core:2.15.2` is hidden in `mvn dependency:tree`, but the SBOM report contains the relationship. As for your concern regarding version overriding, I guess `jackson-core@2.15.0` is overridden by `jackson-core@2.15.2`, and the SBOM describes `jackson-databind@2.15.0` depends on `jackson-core@2.15.2` instead of `jackson-core@2.15.0`. It makes sense to me.

DmitriyLewen replied:
Changed in https://github.com/aquasecurity/go-dep-parser/pull/243/commits/a2af0ef8efec7344712adbbb30c8574b38875918

DmitriyLewen commented:
> I pushed refactoring, but please correct me if I'm wrong.

Idea of using ID for `uniqDeps` looks like good.
I was worried that we would add 2 packages with different versions in "Dependencies".
But I played around with changes and couldn't reproduce my idea.
I think we can use it.

DmitriyLewen commented:
@knqyf263 
I updated PR. Take a look, please.

About using `ID` for `uniqDeps`:
I was able to create case when we use 2 dependencies with different versions. Take a look this test - https://github.com/aquasecurity/go-dep-parser/actions/runs/5996285290/job/16260575992#step:4:42.
That is why i reverted your changes.

knqyf263 commented:
Thanks. I understand why it didn't really make sense to me. The dependencies could be different depending on the version, and it looked weird to overwrite them with the name. We can store all the dependencies and pick up the necessary ones. Would you review my change?

DmitriyLewen commented:
Nice catch! I didn't think about using `libs`. 
I moved convert of `deps` to `libs` loop.
